delivery_schenker/models/delivery_carrier.py

```diff
@@ -0,0 +1,543 @@
+            vals.update({"tracking_number": False, "exact_price": 0})
+            try:
+                response = schenker_request._send_shipping(
+                    vals, self.schenker_booking_type
+                )
+            except Exception as e:
+                raise (e)
+            finally:
+                self._schenker_log_request(schenker_request, picking)
+            if not response:
+                result.append(vals)
+                continue
+            vals["tracking_number"] = response.get("booking_id")
+            # We post an extra message in the chatter with the barcode and the
+            # label because there's clean way to override the one sent by core.
+            body = _("Schenker Shipping barcode document")
+            attachment = []
+            if response.get("barcode"):
+                attachment = [
+                    (
+                        "schenker_label_{}.pdf".format(response.get("booking_id")),
+                        response.get("barcode"),
+                    )
+                ]
+            picking.message_post(body=body, attachments=attachment)
+            result.append(vals)
+        return result
+
+    def schenker_cancel_shipment(self, pickings):
+        """Cancel the expedition
+        :param pickings - stock.picking recordset
+        :returns pdf file
+        """
+        schenker_request = SchenkerRequest(**self._get_schenker_credentials())
+        for picking in pickings.filtered("carrier_tracking_ref"):
+            try:
+                schenker_request._cancel_shipment(picking.carrier_tracking_ref)
+            except Exception as e:
+                raise (e)
+            finally:
+                self._schenker_log_request(schenker_request, picking)
+        return True
+
+    def schenker_get_label(self, reference, reference_type):
+        """Generate label for picking
+        :param picking - stock.picking record
+        :returns pdf file
+        """
+        self.ensure_one()
+        if not reference:
+            return False
+        schenker_request = SchenkerRequest(**self._get_schenker_credentials())
+        label = schenker_request._shipping_label(reference, reference_type)
+        if not label:
+            return False
+        return label
+
+    def schenker_get_tracking_link(self, picking):
+        """Provide tracking link for the customer"""
+        return (
+            "https://eschenker.dbschenker.com/app/tracking-public/?refNumber=%s"
+            % picking.carrier_tracking_ref
+        )
+
+    def schenker_tracking_state_update(self, picking):
+        """Tracking state update"""
+        # TODO: To be implemented
+        self.ensure_one()
+        # if not picking.carrier_tracking_ref:
+        #     return
+        # # schenker_request = SchenkerRequest(
+        # #     **self._get_schenker_credentials(), service="tracking"
+        # # )
+        # Get response
+        return
```

Comment:
This is pending or cannot be done through the API?

Reply:
Thanks @hildickethan ! Yes, we got the docs, but as Schenker doesn't work in the test enviroment with the test API (there's no demo data available to test the responses) we left it until having production data. No it's in a lower priority queue, but if you already have the proper implementation, we can merge this already and cherry pick as you say!